src/main/java/org/opentripplanner/index/IndexGraphQLSchema.java

```diff
@@ -3054,6 +3055,55 @@ private Object getObject(String idString) {
+                                .name("serviceDate")
+                                .description("Date for which the TripTimes are returned. Format: YYYYMMDD. Default: returns all cancelled TripTimes.")
+                                .type(Scalars.GraphQLString)
+                                .build())
+                        .description("Get cancelled TripTimes.")
+                        .type(new GraphQLList(stoptimeType))
+                        .dataFetcher(environment -> {
+                            final String feedId = environment.getArgument("feedId");
+                            ServiceDate maybeServiceDate = null;
+                            try {
+                                maybeServiceDate = ServiceDate.parseString(environment.getArgument("serviceDate"));
+                            } catch (Exception e) {
+
+                            }
+                            final ServiceDate serviceDate = maybeServiceDate;
+                            final TimetableSnapshot snapshot = (index.graph.timetableSnapshotSource != null) ? index.graph.timetableSnapshotSource.getTimetableSnapshot() : null;
+                            final ConcurrentHashMap<TripPattern, Collection<Timetable>> timetableForPattern = new ConcurrentHashMap<>();
+                            return index.tripsForFeedId.get(feedId)
+                                    .stream()
+                                    .flatMap(trip -> {
+                                        final TripPattern pattern = index.patternForTrip.get(trip);
+                                        final Collection<Timetable> timetables = timetableForPattern.computeIfAbsent(pattern, p -> (snapshot != null) ? snapshot.getTimetables(p) : null);
+                                        return ((timetables != null) ? timetables : Arrays.asList(pattern.scheduledTimetable)).stream();
+                                    })
+                                    .filter(timetable -> (serviceDate != null) ? serviceDate.equals(timetable.serviceDate) : timetable.serviceDate != null)
+                                    .flatMap(timetable -> timetable.tripTimes
+                                            .stream()
+                                            .filter(tripTimes -> tripTimes.getRealTimeState() == RealTimeState.CANCELED)
+                                            .map(tripTimes -> {
+                                                final int stopIndex = 0;
+                                                final String agencyId = tripTimes.trip.getId().getAgencyId();
+                                                final Stop stop = timetable.pattern.getStop(stopIndex);
+                                                final CalendarService calendarService = index.graph.getCalendarService();
+                                                final ServiceDay serviceDay = new ServiceDay(index.graph, timetable.serviceDate, calendarService, agencyId);
```

Comment:
Creating unnecessary `ServiceDay` objects, which is slow

Reply:
Nice improvement